emitter.go

```diff
@@ -71,110 +71,123 @@ func CopyMulty(src io.Reader, writers ...io.Writer) (err error) {
+						body = modifier.Rewrite(body)
+
+						// If modifier tells to skip request
+						if len(body) == 0 {
+							filteredRequests[requestID] = time.Now()
+							continue
+						}
+
+						if originalBodyLen != len(body) {
+							payload = append(payload[:headSize], body...)
+						}
+
+						if Settings.debug {
+							Debug("[EMITTER] Rewritten input:", len(payload), "First 500 bytes:", string(payload[0:_maxN]))
+						}
+					} else {
+						if _, ok := filteredRequests[requestID]; ok {
+							delete(filteredRequests, requestID)
+							continue
+						}
 					}
+				}
 
-					if Settings.debug {
-						Debug("[EMITTER] Rewritten input:", len(payload), "First 500 bytes:", string(payload[0:_maxN]))
-					}
-				} else {
-					if _, ok := filteredRequests[requestID]; ok {
-						delete(filteredRequests, requestID)
+				if Settings.prettifyHTTP {
+					payload = prettifyHTTP(payload)
+					if len(payload) == 0 {
 						continue
 					}
 				}
-			}
-
-			if Settings.prettifyHTTP {
-				payload = prettifyHTTP(payload)
-				if len(payload) == 0 {
-					continue
-				}
-			}
 
-			if Settings.splitOutput {
-				// Simple round robin
-				if _, err := writers[wIndex].Write(payload); err != nil {
-					return err
-				}
+				if Settings.splitOutput {
+					// Simple round robin
+					if _, err := writers[wIndex].Write(payload); err != nil {
+						break Loop
+					}
 
-				wIndex++
+					wIndex++
 
-				if wIndex >= len(writers) {
-					wIndex = 0
-				}
-			} else {
-				for _, dst := range writers {
-					if _, err := dst.Write(payload); err != nil {
-						return err
+					if wIndex >= len(writers) {
+						wIndex = 0
+					}
+				} else {
+					for _, dst := range writers {
+						if _, err := dst.Write(payload); err != nil {
+							break Loop
+						}
 					}
 				}
+			} else if nr > 0 {
+				log.Println("WARN: Packet", nr, "bytes is too large to process. Consider increasing --copy-buffer-size")
 			}
-		} else if nr > 0 {
-			log.Println("WARN: Packet", nr, "bytes is too large to process. Consider increasing --copy-buffer-size")
-		}
 
-		// Run GC on each 1000 request
-		if i%1000 == 0 {
-			// Clean up filtered requests for which we didn't get a response to filter
-			now := time.Now()
-			if now.Sub(filteredRequestsLastCleanTime) > 60*time.Second {
-				for k, v := range filteredRequests {
-					if now.Sub(v) > 60*time.Second {
-						delete(filteredRequests, k)
+			// Run GC on each 1000 request
+			if i%1000 == 0 {
+				// Clean up filtered requests for which we didn't get a response to filter
+				now := time.Now()
+				if now.Sub(filteredRequestsLastCleanTime) > 60*time.Second {
+					for k, v := range filteredRequests {
+						if now.Sub(v) > 60*time.Second {
+							delete(filteredRequests, k)
+						}
 					}
+					filteredRequestsLastCleanTime = time.Now()
 				}
-				filteredRequestsLastCleanTime = time.Now()
 			}
-		}
 
-		i++
+			i++
+		}
+	}
+	if er != nil {
+		closeOnce.Do(func() {
+			close(stop)
+		})
```

Comment:
I think this is not a go-way. We should close chan only once a time.

Reply:
Yes. If we close the channel multiple times then it will panic. Previous code would have caused panic if multiple readers or writers fail. sync.Once() is the way to achieve that channel is closed only once.

Reply:
May be we should close channel in other place?

Reply:
Close() explicitly closes the channel and is used by tests only. But I am open to suggestions if you want to call close in some other place.